### DISCUSSION TITLE
scipy/spatial/distance.py warning

### DISCUSSION THREAD

**Discussion by apiszcz** [dstl/Stone-Soup#419] [Q&A]
This started to appear in recent testing.
It may be caused by straight line trajectories, not sure yet.
Any suggestions on locating the root cause are welcome.

```
scipy\spatial\distance.py:1093: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(m)
```

next level down
`lib\scipy\spatial\distance.py(1094)mahalanobis()`

```
    u = _validate_vector(u)
    v = _validate_vector(v)
    VI = np.atleast_2d(VI)
    delta = u - v
    m = np.dot(np.dot(delta, VI), delta)
```
(Pdb++) delta
array([-52882.94101612,   -106.38527496])
(Pdb++) VI
CovarianceMatrix([[1.37098147e-07, 0.00000000e+00],
                  [0.00000000e+00, 1.37098147e-07]])

ss/measures.py/:167
added 
print('u',u,'\nv',v,'\nvi',vi)
```
u [[-106154.26642047]
 [ -55671.58423847]]
v [[-106154.26642047]
 [ -55671.58423847]]
vi [[-1.5479931e-09 -0.0000000e+00]
 [-0.0000000e+00 -1.5479931e-09]]
lib\scipy\spatial\distance.py:1093: RuntimeWarning: invalid value encountered in sqrt
```

**Comment by apiszcz** [comment]
This issue / warning above is still a factor.

**Comment by apiszcz** [comment]
should the error above indicate it is raised by measures.py?

**Comment by sdhiscocks** [comment]
Could you share the value of the covariance (`cov` in `stonesoup.measures.Mahalanobis.__call__`)?

**Comment by apiszcz** [comment]
This is probably caused by my data:

```
        vi = np.linalg.inv(cov)
        print("cov\n",cov)
        return distance.mahalanobis(u, v, vi)
```

```
cov
 [[-1.29313979e+12  0.00000000e+00]
 [ 0.00000000e+00 -1.29313979e+12]]
C:\g\via37\lib\scipy\spatial\distance.py:1093: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(m)
cov
 [[-1.29313979e+12  0.00000000e+00]
 [ 0.00000000e+00 -1.29313979e+12]]
cov
 [[-1.26821863e+12  0.00000000e+00]
 [ 0.00000000e+00 -1.26821863e+12]]
```

**Comment by sdhiscocks** [comment]
Thanks. Looks like issue is negative covariance matrix? 

Does setting `force_symmetric_covariance` on [KalmanUpdater](https://stonesoup.readthedocs.io/en/v0.1b3/stonesoup.updater.html#stonesoup.updater.kalman.ExtendedKalmanUpdater) help?

Any thoughts @jmbarr?

**Comment by apiszcz** [comment]
Thank you for the suggestion:

Tested the following with no difference still returning 
lib\scipy\spatial\distance.py:1093: RuntimeWarning: invalid value encountered in sqrt

```
        #updater           = KalmanUpdater(measurement_model)
        updater           = ExtendedKalmanUpdater(measurement_model,force_symmetric_covariance=True)
```

**Comment by jmbarr** [comment]
Yup - that can't be a valid covariance matrix. The elements on the diagonal are the variance of the elements of the state vector and so have to be positive. So something went wrong when calculating `cov`. The `force_symmetric` flag won't work as it will just ensure that the matrix is symmetric, not positive semidefinite.

This sometimes happens when the timings get mangled and you end up trying to 'predict' into the past. Check the timestamps perhaps?

**Comment by apiszcz** [comment]
Ok, that is a possibility (timestamp issue), this happens when I'm out of new data. Will review. Thanks.

>>
Reviewed timestamps when warning/error appears, monotonic and have 1 to 70 seconds of separation.

**Comment by jmbarr** [comment]
Can you trace back to the operation that caused `cov` diagonal elements to go negative? Might be the result of an operation near/over the floating-point limit..
